plugin/pkg/scheduler/schedulercache/cache.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schedulercache
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+var cleanAssumedPeriod = 1 * time.Second
+
+// New returns a Cache implementation.
+// It automatically starts a go routine that manages expiration of assumed pods.
+// "ttl" is how long the assumed pod will get expired.
+// "stop" is the channel that would close the background goroutine.
+func New(ttl time.Duration, stop chan struct{}) Cache {
+	cache := newSchedulerCache(ttl, cleanAssumedPeriod, stop)
+	cache.run()
+	return cache
+}
+
+type schedulerCache struct {
+	stop   chan struct{}
+	ttl    time.Duration
+	period time.Duration
+
+	// This mutex guards all fields within this cache struct.
+	mu          sync.Mutex
+	assumedPods map[string]assumedPod
+	podStates   map[string]podState
```

Comment:
Moving from parent PR discussion:

@wojtekt:
"I think it would be cleaner to have
pods map[string]podInfo

podInfo:
pod *api.Pod
state podState
deadline time.Time

that to have two maps (that are slightly duplicating the information). WDYT?
 @wojtek-t  kubernetes member
wojtek-t added a note 22 hours ago
Or even we will not need state in such case, because we can define it as:
deadline = 0 <=> pod is added
deadline != 0 <=> pod is assumed
WDYT?"

@hongchaodeng :

""assumedPods" is expected to be a much smaller subset of pods. It's separated out to do efficient scanning and cleaning up."